runtime/lua/vim/lsp/completion.lua

```diff
@@ -463,6 +463,188 @@ local function request(clients, bufnr, win, ctx, callback)
+      cmp_info
+  end
+
+  return function(bufnr, param, select_word)
+    cleanup()
+
+    infoContext.bufnr = bufnr
+    infoContext.word = select_word
+    local debounce_time = next_doc_debounce()
+
+    infoContext.timer = assert(vim.uv.new_timer())
+    infoContext.timer:start(
+      debounce_time,
+      0,
+      vim.schedule_wrap(function()
+        local valid, cmp_info = is_valid()
+        if not valid then
+          cleanup()
+          return
+        end
+
+        cancel_pending()
+        local client_id = vim.tbl_get(cmp_info.completed, 'user_data', 'nvim', 'lsp', 'client_id')
+        local client = client_id and lsp.get_client_by_id(client_id)
+        if not client then
+          return
+        end
+
+        local start_time = vim.uv.hrtime()
+        infoContext.last_request_time = start_time
+
+        local ok, request_id = client:request(
+          vim.lsp.protocol.Methods.completionItem_resolve,
+          param,
+          function(err, result)
+            local end_time = vim.uv.hrtime()
+            local response_time = (end_time - start_time) * ns_to_ms
+            doc_rtt_ms = doc_compute_new_average(response_time)
+
+            if err or not result or next(result) == nil then
+              if err then
+                vim.notify(err.message, vim.log.levels.WARN)
+              end
+              return
+            end
+
+            valid, cmp_info = is_valid()
+            if not valid then
+              return
+            end
+
+            local value = vim.tbl_get(result, 'documentation', 'value')
+            if not value then
+              return
+            end
+
+            local windata = api.nvim__complete_set(cmp_info.selected, {
+              info = value,
+            })
+            local kind = vim.tbl_get(result, 'documentation', 'kind')
+            update_popup_window(windata.winid, windata.bufnr, kind)
+          end,
+          bufnr
+        )
+
+        if ok and request_id then
+          infoContext.request_ids[client.id] = request_id
+        end
+      end)
+    )
+  end
+end
+
+local debounce_info_request = cmpitem_resolve()
+
+local function on_completechanged(group, bufnr)
+  api.nvim_create_autocmd('CompleteChanged', {
+    group = group,
+    buffer = bufnr,
+    callback = function(args)
+      local completed_item = api.nvim_get_vvar('event').completed_item or {}
+      if (completed_item.info or '') ~= '' then
+        local data = vim.fn.complete_info({ 'selected' })
+        update_popup_window(data.preview_winid, data.preview_bufnr)
+        return
+      end
+
+      if
+        #lsp.get_clients({
+          id = vim.tbl_get(completed_item, 'user_data', 'nvim', 'lsp', 'client_id'),
+          method = ms.completionItem_resolve,
+          bufnr = args.buf,
+        }) == 0
+      then
+        return
+      end
+
+      local param = vim.tbl_get(completed_item, 'user_data', 'nvim', 'lsp', 'completion_item')
```

Comment:
where is this user_data coming from? a 1-line comment should give a hint.

there is way too much "hidden knowledge" in this PR. use concise 1-line comments to give hints.

Reply:
https://github.com/neovim/neovim/blob/06613989a67e03044941a608536a775af1521c61/runtime/lua/vim/lsp/completion.lua#L338-L342

Reply:
added comment